lib/teleterm/apiserver/handler/handler_unified_resources.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handler
+
+import (
+	"context"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	api "github.com/gravitational/teleport/gen/proto/go/teleport/lib/teleterm/v1"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+)
+
+var supportedResourceKinds = []string{
+	types.KindNode,
+	types.KindDatabase,
+	types.KindKubernetesCluster,
+}
+
+func (s *Handler) ListUnifiedResources(ctx context.Context, req *api.ListUnifiedResourcesRequest) (*api.ListUnifiedResourcesResponse, error) {
+	clusterURI, err := uri.Parse(req.GetClusterUri())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sortBy := types.SortBy{}
+	if req.GetSortBy() != nil {
+		sortBy.IsDesc = req.GetSortBy().IsDesc
+		sortBy.Field = req.GetSortBy().Field
+	}
+
+	kinds := req.GetKinds()
+	if len(kinds) == 0 {
+		kinds = supportedResourceKinds
+	} else {
+		for _, kind := range kinds {
+			if !slices.Contains(supportedResourceKinds, kind) {
+				return nil, trace.BadParameter("unsupported resource kind: %s", kind)
+			}
+		}
+	}
```

Comment:
I don't think the handler should be validating resource kinds. What do you think about moving this to `unifiedresources.List`? This way the core logic related to this feature stays in a dedicated package. With the current implementation, some of it is expressed in `unifiedresources` and some in `handler`.

See https://github.com/gravitational/teleport/issues/14011.

Reply:
I had some troubles with deciding what should go where.
For example, we are parsing cluster URI in the handler to prevent malformed data entering the deeper layer. It seemed to me that validating `kinds` is a similar thing - it prevents unsupported kinds from being passed down. 

OTOH now that I think of it, there is a difference in checking if data has the correct format vs validating the actual values, so I think it makes sense to move it to `unifiedresources`.